### PR TITLE
WIP Media widget, providing access to the camera

### PR DESCRIPTION
**Pull request by jasongrout**
This is still very much a work-in-progress:
- [ ] Expose the full MediaConstraints api
- [ ] figure out why it's slow
- [ ] provide a trigger to take a screenshot
- [ ] is there a way to record video/audio?

@SylvainCorlay :)

### CODE REVIEW

**Comment by SylvainCorlay** [comment]
👍

**Comment by jasongrout** [comment]
By the way, Sylvain and I discovered it is "slow" in that it only seems like the video preview is working when the browser repaints (e.g., scrolling, or changing window size).

**Comment by dmadeka** [comment]
@jasongrout Ideally, we should also have the ability to transfer the stream at some time interval over to the python side? (if the record is done, then analysis can be done directly with Python for NLP type stuff right?)

File: jupyter-js-widgets/src/widget_media.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    initialize: function() {
+        // Get the camera permissions
+        this.stream = navigator.mediaDevices.getUserMedia({audio: false, video: true});
+    }
+});
+
+var MediaView = widget.DOMWidgetView.extend({
+
+    initialize: function() {
+        this.setElement(document.createElement('video'));
```
Comment (dmadeka) [inline jupyter-js-widgets/src/widget_media.js]: `this.el.autoplay = true`

to stop the 'slow' rendering

**Comment by jasongrout** [comment]
@dmadeka - do you want to take these commits and finish it up to whatever looks good to you and submit it as a separate PR?

**Comment by gnestor** [comment]
@jasongrout Let me know how I can help with this one...

**Comment by dmadeka** [comment]
@gnestor I got a little far with this actually. So, the question is what do you send back to the python side? I was serializing the model using the `Byte` traitlet. But I think this is totally inefficient to continuously sync the stream. It seems to require something intelligent to know when to send the stream back to the python side, or might need something eventful to send just the diff..

if we are able to send the stream back though, we can use the voice-recognition/nlp tools of Python to enable voice-activated applications in the notebook!

**Comment by dmadeka** [comment]
@gnestor At least that was my initial goal, not sure what @jasongrout or @SylvainCorlay envision for the widget

**Comment by gnestor** [comment]
@dmadeka Would it be possible or efficient to just write the stream to the filesystem and relay the file path and stream status back to IPython?

**Comment by dmadeka** [comment]
@gnestor I haven't tried that actually, should the user specify the path in Python and then the JS side sends a custom message on some condition for the stream? 

By the way, I know the stream is written to a DOM File, maybe we can just access that? Im not too sure about how that works

**Comment by gnestor** [comment]
Ya I'm not sure either. I think that file path can be an optional argument for the Media widget and if none is given, it can default to a temp directory. If the stream is already being written somewhere and we can access that file path, even better!

**Comment by jasongrout** [comment]
We could take this a lot of directions, like both of you mention. How about for now, we just have an event that will take a snapshot and save that image as a png data url? I think that's a decent minimally viable widget.

Unfortunately, now the code needs to be updated to typescript, but the current widget code should provide a good template.

As for recording to the filesystem, that gets tricky. Very tricky. The widget doesn't have access to the filesystem itself, and doesn't really have access to the notebook server. Instead, it only has access to the kernel through a comm channel. 

I suppose we could record to a buffer and then periodically sync that buffer to the kernel. We'd probably need a custom serializer that would only transmit new data.

**Comment by gnestor** [comment]
@jasongrout Ok so it sounds like there is no precedent for a widget like this? In that case, I agree that it would behoove us to go with a minimum viable widget first (take a still screenshot) . I can update to Typescript and try to implement this, unless you would like to see this through.

**Comment by jasongrout** [comment]
Go for it. I'm happy to let you do this.

I've noticed you've been doing a lot on widgets lately. Thanks!

**Comment by gnestor** [comment]
Ok, cool! This will be a good learning experience! I'm happy to help!

**Comment by SylvainCorlay** [comment]
@dmadeka do you want to complete this for 6.0?

**Comment by maartenbreddels** [comment]
I think so, we can let it evolve in ipywebrtc faster than here.

**Pull request closed without merging**